### PR TITLE
Data Layer: Allow Redux actions as network request handlers

### PR DESCRIPTION
**Pull request by dmsnell**
Pretty often I see network requests made via `dispatchRequest()` and the
response handler for `onFailure` is nothing more than a notice to alert
our customers that something failed.

```js
const announceFailure = ( { dispatch } ) => errorNotice( translate(
'Sad' ) );
```

Well this is a lot of code. In many cases we just want to dispatch an
action. This is a simple thing and we should not require a function
where a simple object will do.

```js
dispatchRequest(
	fetchThing,
	storeThingInState,
	errorNotice( translate( 'Sad' ) ),
)
```

Actually there's one major flaw here: translations need to happen at the
time of dispatch, not at app boot as they would here, so that's a thing
to resolve.

Naively we could wrap `dispatchRequest()` but I think this is a terrible
idea.

### CODE REVIEW

File: client/state/data-layer/wpcom-http/README.md [hunk 1]
```diff
@@ -286,6 +286,25 @@ export default {
...
+One of the most common types of responders is the `announceFailure()` handler which does nothing more dispatch a single Redux action.
+Creating multiple functions just to return a static object can be cumbersome.
+Let's take advantage of a shortcut designed just for these situations: we can choose to send in a disptchable object in place of the handler functions.
+
+```js
+export default {
+	[ REQUEST_LIKES ]: [ dispatchRequest(
+		fetchLikes,
+		updateLikes,
+		createNotice( translate( 'Could not fetch likes, try again.' ) ), // <- this is a simple action
```
Comment (gwwar) [inline client/state/data-layer/wpcom-http/README.md]: I like the simplification here. Are we ok here if someone changes their language? I think we still do a full page refresh, but wasn't sure.

Comment (dmsnell) [reply client/state/data-layer/wpcom-http/README.md]: that's my hope, because if we can forego this one constraint it will make so many cases much easier

File: client/state/data-layer/wpcom-http/README.md [hunk 2]
```diff
@@ -329,3 +348,35 @@ export default {
 	) ]
 }
 ```
+
```
Comment (samouri) [inline client/state/data-layer/wpcom-http/README.md]: This documentation change deserves its own PR because it is unrelated to `dispatchRequest` accepting an object

**Comment by samouri** [comment]
I really like this change and it reminds me a bit of the simplification that `redux-connect` allows for its `mapPropsToDispatch` argument where you can provide an object instead of a function.

Lets decide on what to do about https://github.com/Automattic/wp-calypso/pull/18733 vs `apiDump` and then move forward afterwards

**Comment by dmsnell** [comment]
> Lets decide on what to do about #18733 vs apiDump and then move forward afterwards.

I kind of see this change as independent to #18733 and something it _and_ `apiData()` can both benefit from.

Do you feel differently?

File: client/state/data-layer/wpcom-http/README.md (hunk 1 above)
Comment (jsnajdr) [reply client/state/data-layer/wpcom-http/README.md]: > Are we ok here if someone changes their language? I think we still do a full page refresh, but wasn't sure.

We do a full page refresh. I explored the question "do we really need to refresh?" in #13712 as a part of my trial project. TLDR is that it's avoidable and the biggest problem is switching from LTR to RTL languages and swapping `style.css` and `style-rtl.css` stylesheets.

Across Calypso, effort is being made to make sure that `translate` is always called "just in time", on every render, with the latest l10n data. React components use the `localize` HOC that forces a rerender on locale change. "Static" localized strings are implemented with something like this:
```js
// don't do this at global module level
const noticeTexts = { in: i18n.translate( 'Hello' ), out: i18n.translate( 'Goodbye' ) };
// do this instead
const noticeTexts = translate => ( { in: translate( 'Hello' ), out: translate( 'Goodbye' ) } );
```

In this sense, static `createNotice( translate( ... ) )` is a regression, although it doesn't immediately break anything.

One solution is to avoid static global actions and do a trivial action creator instead:
```js
dispatchRequest( ..., () => createNotice( i18n.translate( 'Error' ) ) );
```

Now, when a locale gets switched, new notices use the new language, but the old ones, already displayed, use the old language.

I don't know right now what's the perfect solution. I'm afraid that trying to do UI stuff in a data layer handler will always be a bit challenging.

Comment (dmsnell) [reply client/state/data-layer/wpcom-http/README.md]: This probably needs to pair with changes to the data layer API if we are going to make them return actions, in which case it gets a bit easier.

> I'm afraid that trying to do UI stuff in a data layer handler will always be a bit challenging.

You've raised this issue before but I'd argue there's a difference between doing UI stuff in the data layer and raising global alerts through the notices. In almost every case it's because some network activity failed and by keeping the error-handling out of React we're keeping a considerably simpler model (so many `_FAILURE` Redux actions fly by without even having a single reducer respond let alone UI) and doing a very pragmatic thing: inform the customer that something failed and they might want to try again. If it were actually coupling into the app more tightly and weren't just a basic string I'd be all in with you on that assessment.

**Comment by jsnajdr** [comment]
I like this change -- it's very similar to other improvements in #18733. We can either merge this independently, or update #18733.

Some unit tests would be desirable before merging.

I see only one little trouble: the request handler can now be either a plain object (i.e., an action), or it can be a function (i.e., a handler to be executed, or an action creator in #18733).

What if I supply a thunk as the value? It's a dispatchable action, so it should be passed to `dispatch`, but `dispatch` request will do the wrong thing: it will identify it as function and execute it with wrong parameters.

I don't mind if we decide that this API doesn't support thunks. Just wanted to point out that there is a shady area where `dispatchRequest` can't easily decide what the user intended when she passed certain thing as a handler.

File: client/state/data-layer/wpcom-http/README.md (hunk 1 above)
Comment (dmsnell) [reply client/state/data-layer/wpcom-http/README.md]: Noting that this won't be a problem if it happens after changes to the data layer API found in #18733 - in that case there will be no distinction between functions returning actions in general and data layer handlers

that is, after that change we should just be able to send…

```js
dispatchRequest( {
	onFailure: () => createNotice( translate( 'Error' ) )
} )
```

**Comment by dmsnell** [comment]
Closing because when #18733 lands this should no longer be necessary - we will be able to supply dummy actions creators not based on the originating action. It's a few extra characters of code and an extra function to create, but it's not a hassle and it's much simpler…

```js
dispatchRequest( {
	fetch: () => http( … ),
	onSuccess: addThing,
	onFailure: () => errorNotice( translate( 'Failed to update stuff' ) ),
} )
```
